File: common/src/main/java/juuxel/adorn/client/book/Page.java
package juuxel.adorn.client.book;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import juuxel.adorn.util.EntryOrTag;
import net.minecraft.world.item.Item;
import net.minecraft.world.level.ItemLike;
import net.minecraft.world.item.ItemStack;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.core.registries.Registries;
import net.minecraft.tags.TagKey;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.ComponentSerialization;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

public record Page(List<Icon> icons, Component title, Component text, @Nullable Image image) {
    public static final Codec<Page> CODEC = RecordCodecBuilder.create(instance -> instance.group(
        Icon.CODEC.listOf().fieldOf("icons").forGetter(Page::icons),
        ComponentSerialization.CODEC.fieldOf("title").forGetter(Page::title),
        ComponentSerialization.CODEC.optionalFieldOf("text", Component.empty()).forGetter(Page::text),
        Image.CODEC.optionalFieldOf("image").forGetter(page -> Optional.ofNullable(page.image))
    ).apply(instance, Page::new));

    // For DFU
    private Page(List<Icon> icons, Component title, Component text, Optional<Image> image) {
        this(icons, title, text, image.orElse(null));
    }

    public static Builder builder() {
        return new Builder();
    }

    public static final class Builder {
        private final List<Icon> icons = new ArrayList<>();
        private @Nullable Component title;
        private @Nullable Component text;
        private @Nullable Image image;

        private Builder() {
        }

        public Builder icon(ItemLike item) {
            icons.add(new Icon(new EntryOrTag.OfEntry<>(item.asItem())));
            return this;
        }

        public Builder icon(TagKey<Item> tag) {
            icons.add(new Icon(new EntryOrTag.OfTag<>(tag)));
            return this;
        }

        public Builder title(Component title) {
            this.title = title;
            return this;
        }

        public Builder content(Component text) {
            this.text = text;
            return this;
        }

        public Builder image(Image image) {
            this.image = image;
            return this;
        }

        public Page build() {
            if (icons.isEmpty()) throw new IllegalArgumentException("Page has no icons");
            Objects.requireNonNull(title);
            Objects.requireNonNull(text);
            return new Page(icons, title, text, image);
        }
    }

    public record Icon(EntryOrTag<Item> items) {
        public static final Codec<Icon> CODEC = EntryOrTag.codec(Registries.ITEM).xmap(Icon::new, Icon::items);

        public List<ItemStack> createStacks() {
            return switch (items) {
                case EntryOrTag.OfEntry(var item) -> List.of(item.getDefaultInstance());
                case EntryOrTag.OfTag(var tag) -> BuiltInRegistries.ITEM.get(tag).map(entries -> {
                    List<ItemStack> result = new ArrayList<>(entries.size());
                    for (var entry : entries) {
                        result.add(entry.value().getDefaultInstance());
                    }
                    return result;
                }).orElse(List.of());
            };
        }
    }
}

File: common/src/main/kotlin/juuxel/adorn/client/book/Page.kt
package juuxel.adorn.client.book

import com.mojang.datafixers.util.Pair
import com.mojang.serialization.Codec
import com.mojang.serialization.DataResult
import com.mojang.serialization.DynamicOps
import com.mojang.serialization.codecs.RecordCodecBuilder
import juuxel.adorn.util.MoreCodecs
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.registry.Registries
import net.minecraft.registry.RegistryKeys
import net.minecraft.registry.tag.TagKey
import net.minecraft.text.Text
import net.minecraft.util.Identifier
import java.util.Optional

data class Page(
    val icons: List<Icon>,
    val title: Text,
    val text: Text,
    val image: Image?
) {
    // For DFU
    private constructor(icons: List<Icon>, title: Text, text: Text, image: Optional<Image>) :
        this(icons, title, text, image.orElse(null))

    companion object {
        val CODEC: Codec<Page> = RecordCodecBuilder.create { instance ->
            instance.group(
                Icon.CODEC.listOf().fieldOf("icons").forGetter { it.icons },
                MoreCodecs.TEXT.fieldOf("title").forGetter { it.title },
                MoreCodecs.TEXT.optionalFieldOf("text", Text.empty()).forGetter { it.text },
                Image.CODEC.optionalFieldOf("image").forGetter { Optional.ofNullable(it.image) }
            ).apply(instance, ::Page)
        }
    }

    sealed class Icon {
        abstract fun createStacks(): List<ItemStack>

        data class ItemIcon(val item: Item) : Icon() {
            override fun createStacks(): List<ItemStack> =
                listOf(item.defaultStack)
        }

        data class TagIcon(val tag: TagKey<Item>) : Icon() {
            override fun createStacks(): List<ItemStack> =
                Registries.ITEM.getOrCreateEntryList(tag).map { it.value().defaultStack }
        }

        companion object {
            val CODEC: Codec<Icon> = object : Codec<Icon> {
                override fun <T> encode(input: Icon, ops: DynamicOps<T>, prefix: T): DataResult<T> {
                    val id = when (input) {
                        is ItemIcon -> Registries.ITEM.getId(input.item).toString()
                        is TagIcon -> "#${input.tag.id}"
                    }
                    return ops.mergeToPrimitive(prefix, ops.createString(id))
                }

                override fun <T> decode(ops: DynamicOps<T>, input: T): DataResult<Pair<Icon, T>> =
                    ops.getStringValue(input).map { id ->
                        if (id.startsWith('#')) {
                            TagIcon(TagKey.of(RegistryKeys.ITEM, Identifier(id.substring(1))))
                        } else {
                            ItemIcon(Registries.ITEM[Identifier(id)])
                        }
                    }.map { Pair.of(it, ops.empty()) }
            }
        }
    }
}
